changefeedccl: allow users to alter the sink URI of an existing changefeed

sherman-grewal opened the pull request:
changefeedccl: allow users to alter the sink URI of
an existing changefeed

References #75895

In this PR, we introduce the capability to alter
the sink URI of an existing changefeed. This
can be achieved by executing the following
statement:

ALTER CHANGEFEED <job_id> SET sink = '<sink_uri>'

Note that the sink type cannot be altered. That is,
the sink type must be the same type that was chosen
when the changefeed was initially created.

Release note (enterprise change): Users may now alter
the sink URI of an existing changefeed. This can be
achieved by executing the following statement:

ALTER CHANGEFEED <job_id> SET sink = '<sink_uri>'

Where the sink type of the new sink must match the
sink type of the old sink that was chosen at the
creation of the changefeed.

miretskiy reviewed (approved):
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77043)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77043#-MwmO2jxBDer43CRMFwe:-MwmO2jxBDer43CRMFwf:beo3xi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec6ade47bd5eacf5616fb165e94ffcc92bf1d7d8/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L181)):*
> ```Go
> 					}
> 					if key == changefeedbase.OptSink {
> 						newChangefeedStmt.SinkURI = tree.NewDString(value)
> ```

This PR is very nice; consider augmenting this logic so that if 
newURI scheme differs from old url scheme, then you wipe all of the previously specified sink specific options.


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: allow users to alter the sink URI of
an existing changefeed

In this PR, we introduce the capability to alter
the sink URI of an existing changefeed. This
can be achieved by executing the following
statement:

ALTER CHANGEFEED <job_id> SET sink = '<sink_uri>'

Note that the sink type cannot be altered. That is,
the sink type must be the same type that was chosen
when the changefeed was initially created.

Release note (enterprise change): Users may now alter
the sink URI of an existing changefeed. This can be
achieved by executing the following statement:

ALTER CHANGEFEED <job_id> SET sink = '<sink_uri>'

Where the sink type of the new sink must match the
sink type of the old sink that was chosen at the
creation of the changefeed.

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77043)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77043#-MwmO2jxBDer43CRMFwe:-Mwmj7KG6BcuhkXV0Rb5:b-ucqiiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec6ade47bd5eacf5616fb165e94ffcc92bf1d7d8/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L181)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This PR is very nice; consider augmenting this logic so that if 
newURI scheme differs from old url scheme, then you wipe all of the previously specified sink specific options.

</blockquote></details>

As discussed offline, we will be disallowing users to alter changefeed schemes for now. When this option becomes available to users, we will keep this in mind.


<!-- Sent from Reviewable.io -->

sherman-grewal commented:
bors r+

shermanCRL commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -174,21 +175,44 @@ func alterChangefeedPlanHook(
+						if newSinkURI.Scheme != prevSinkURI.Scheme {
+							return pgerror.Newf(
+								pgcode.InvalidParameterValue,
+								`new sink type %q does not match original sink type %q, sink type cannot be altered`,
```
It’d be nice to make it a little more prescriptive, so the user knows what to do next. something like ‘changing the type of a changefeed (in this case %q to %q) is disallowed. use CREATE CHANGEFEED instead.'
